### PR TITLE
[CARBONDATA-198] System level lock feature for compaction.

### PR DESCRIPTION
**Pull request by ravikiran23**
1. implemented system level compaction lock feature.
2. made the DDL call to compaction as blocking call.
3. handled the requests which are received during one compaction and taking those requests later.

### CODE REVIEW

File: integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala [hunk 1]
```diff
@@ -344,49 +540,77 @@ object CarbonDataRDDFactory extends Logging {
...
+              executor, sqlContext, kettleHomePath, storeLocation
+            )
+            // check for all the tables.
+            val isSystemCompactionLockEnabled = CarbonProperties.getInstance()
+              .getProperty(CarbonCommonConstants.ENABLE_SYSTEM_LEVEL_COMPACTION_LOCK, "false")
+              .equalsIgnoreCase("true")
+
+            if (isSystemCompactionLockEnabled) {
+              logger.info("System level compaction lock is enabled.")
+              CarbonEnv.getInstance(sqlContext).carbonCatalog.metadata.tablesMeta.map { c =>
```
Comment (gvramana) [inline integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala]: provide getnexttabletocompact function

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java [hunk 2]
```diff
@@ -904,6 +904,27 @@
...
+  /**
+   * File created in case of major compaction request
+   */
+  public static String majorCompactionRequiredFile = "compactionRequired_major";
+
+  /**
+   * Property for enabling system level compaction lock.1 compaction can run at once.
+   */
+  public static String ENABLE_SYSTEM_LEVEL_COMPACTION_LOCK =
+      "carbon.enable.systemlevel.compaction.lock";
```
Comment (gvramana) [inline core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java]: change configuration to "carbon.concurrent.compaction" default to false. i.e., no concurrent compaction allowed by default.

Comment (ravikiran23) [inline core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java]: fixed.

File: integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala (hunk 1 above)
Comment (ravikiran23) [inline integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala]: done

File: integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala [hunk 3]
```diff
@@ -297,15 +366,144 @@ object CarbonDataRDDFactory extends Logging {
...
+        .audit("Not able to acquire the system level compaction lock for table " +
           s"${carbonLoadModel.getDatabaseName}.${carbonLoadModel.getTableName}"
         )
       logger
         .error("Not able to acquire the compaction lock for table " + carbonLoadModel
           .getDatabaseName + "." + carbonLoadModel.getTableName
         )
-      sys.error("Table is already locked for compaction. Please try after some time.")
+      if (!CarbonCompactionUtil
+        .createCompactionRequiredFile(carbonTable.getMetaDataFilepath, compactionType)) {
```
Comment (gvramana) [inline integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala]: move printing log to  function

File: integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala [hunk 4]
```diff
@@ -333,7 +531,7 @@ object CarbonDataRDDFactory extends Logging {
           )
     }
 
-    var loadsToMerge = CarbonDataMergerUtil.identifySegmentsToBeMerged(
+    val loadsToMerge = CarbonDataMergerUtil.identifySegmentsToBeMerged(
```
Comment (gvramana) [inline integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala]: this condition is not required before thread launch

File: integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala [hunk 5]
```diff
@@ -400,43 +631,36 @@ object CarbonDataRDDFactory extends Logging {
...
+        // making this an blocking call for DDL
+        compactionThread.run()
+      }
+      else {
+        // non blocking call in case of auto compaction.
+        compactionThread.start()
+      }
     }
     else {
       compactionLock.unlock()
```
Comment (gvramana) [inline integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala]: check if any exception is possible , if needed to be moved to finally

File: integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala (hunk 3 above)
Comment (ravikiran23) [inline integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala]: done

File: integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala [hunk 6]
```diff
@@ -297,15 +366,144 @@ object CarbonDataRDDFactory extends Logging {
...
+          .getDatabaseName + "." + carbonLoadModel.getTableName
+        )
+      }
+      else {
+        logger
+          .info("successfully created a compaction required file for table " + carbonLoadModel
+            .getDatabaseName + "." + carbonLoadModel.getTableName
+          )
+      }
+      sys.error("System is already locked for compaction. Please try after some time.")
```
Comment (ravikiran23) [inline integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala]: done

File: integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala (hunk 4 above)
Comment (ravikiran23) [inline integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala]: fixed

File: integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala (hunk 5 above)
Comment (ravikiran23) [inline integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala]: fixed

**Pull request closed without merging**